test/core/resource_quota/arena_test.cc

```diff
@@ -0,0 +1,171 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "src/core/lib/resource_quota/arena.h"
+
+#include <inttypes.h>
+#include <string.h>
+
+#include <algorithm>
+#include <ostream>
+#include <string>
+#include <vector>
+
+#include "absl/memory/memory.h"
+#include "absl/strings/str_join.h"
+#include "gtest/gtest.h"
+
+#include <grpc/grpc.h>
+#include <grpc/support/sync.h>
+#include <grpc/support/time.h>
+
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/core/lib/gprpp/thd.h"
+#include "src/core/lib/resource_quota/resource_quota.h"
+
+using grpc_core::Arena;
+
+static auto* g_memory_allocator = new grpc_core::MemoryAllocator(
+    grpc_core::ResourceQuota::Default()->memory_quota()->CreateMemoryAllocator(
+        "test"));
+
+TEST(ArenaTest, NoOp) { Arena::Create(1, g_memory_allocator)->Destroy(); }
+
+TEST(ArenaTest, ManagedNew) {
+  Arena* arena = Arena::Create(1, g_memory_allocator);
+  for (int i = 0; i < 100; i++) {
+    arena->ManagedNew<std::unique_ptr<int>>(absl::make_unique<int>(i));
+  }
+  arena->Destroy();
+}
+
+struct AllocShape {
+  size_t initial_size;
+  std::vector<size_t> allocs;
+};
+
+std::ostream& operator<<(std::ostream& out, const AllocShape& shape) {
+  out << "AllocShape{initial_size=" << shape.initial_size
+      << ", allocs=" << absl::StrJoin(shape.allocs, ",") << "}";
+  return out;
+}
+
+class AllocTest : public ::testing::TestWithParam<AllocShape> {};
+
+TEST_P(AllocTest, Works) {
```

Comment:
nit: There's no great candidate for a better name than `Works` here, which I think means that this test could be split up to be more specific. It covers attributes of the allocations themselves, exercises arenas with and without sufficient initial size, and ensures that the same addresses aren't handed out repeatedly.

Reply:
Yeah maybe... I actually think this might be a good candidate for some fuzzing, but not today.